linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,134 @@
+
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, int, completion_port_add, COMPLETION_PORT_HANDLE, completion_port, int, epoll_op, SOCKET_HANDLE, socket,
+    ON_COMPLETION_PORT_EVENT_COMPLETE, event_callback, void*, event_callback_ctx);
+MOCKABLE_FUNCTION(, void, completion_port_remove, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+### completion_port_create
+
+```C
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+```
+
+`completion_port_create` creates and initializes the completion port module
+
+`completion_port_create` shall allocate memory for a completion port object.
+
+`completion_port_create` shall create the epoll instance by calling `epoll_create`.
+
+`completion_port_create` shall create a thread that runs `epoll_worker_func` to handle the epoll events.
+
+On success `completion_port_create` shall return the allocated `COMPLETION_PORT_HANDLE`.
+
+If there are any errors then `completion_port_create` shall fail and return `NULL`.
+
+### completion_port_inc_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_inc_ref` handles the incrementing of the reference count of the module
+
+If `completion_port` is `NULL`, `completion_port_inc_ref` shall return.
+
+Otherwise `completion_port_inc_ref` shall increment the internally maintained reference count.
+
+### completion_port_dec_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_dec_ref` handles the decrement of the reference count and freeing the memory if necessary
+
+`completion_port_dec_ref` shall decrement the reference count for `completion_port`.
+
+If the reference count reaches 0, `completion_port_dec_ref` shall do the following:
+
+- increment the flag signaling that the threads can complete.
+
+- close the epoll object.
+
+- close the thread by calling `ThreadAPI_Join`
+
+- then the memory associated with `completion_port` shall be freed.
+
+### completion_port_add
+
+```C
+MOCKABLE_FUNCTION(, int, completion_port_add, COMPLETION_PORT_HANDLE, completion_port, int, epoll_op, SOCKET_HANDLE, socket,
+    ON_COMPLETION_PORT_EVENT_COMPLETE, event_callback, void*, event_callback_ctx);
+```
+
+`completion_port_add` adds an ingress item to the post queue for epoll signaling
+
+If `completion_port` is `NULL`, `completion_port_add` shall return a non-NULL value.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_add` shall return a non-NULL value.
+
+If `event_callback` is `NULL`, `completion_port_add` shall return a non-NULL value.
+
+`completion_port_add` shall allocate a `EPOLL_THREAD_DATA` object to store thread data.
```

Comment:
>ort_add` shall allocate a `EPOLL_THREAD_DATA` object to store thread data. [](http://example.com/codeflow?start=13&length=74)

Ehm, will this trigger a memory allocation?
It would be so awesome to have a queue backed by one of these TARRAYs, wouldn't it?

Reply:
Maybe for another task?
I remember we had a similar thing for the threadpool work items for linux

Reply:
I can create a task for this.